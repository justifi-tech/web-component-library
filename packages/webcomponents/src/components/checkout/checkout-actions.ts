import { CheckoutService } from '../../api/services/checkout.service';
import { getErrorMessage } from '../../api/services/utils';

export const makeGetCheckout = ({ authToken, checkoutId, service }) =>
  async ({ onSuccess, onError }) => {
    try {
      const response = await (service as CheckoutService).fetchCheckout(authToken, checkoutId);

      if (!response.error) {
        const checkout = response.data;
        onSuccess({ checkout });
      } else {
        const responseError = getErrorMessage(response.error);
        return onError(responseError);
      }
    } catch (error) {
      return onError(error.message || error);
    }
  };

export const makePay = ({ authToken, checkoutId, service }) =>
  async ({ paymentMethodToken, onSuccess, onError }) => {
    try {
      const response = await (service as CheckoutService).pay(authToken, checkoutId, paymentMethodToken);

      if (!response.error) {
        const checkout = response.data;
        onSuccess({ checkout });
      } else {
        const responseError = getErrorMessage(response.error);
        return onError(responseError);
      }
    } catch (error) {
      return onError(error.message || error);
    }
  };
