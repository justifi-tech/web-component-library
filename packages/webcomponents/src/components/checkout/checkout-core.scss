@import 'reboot';
@import 'grid';
@import 'buttons';
@import 'spinners';
@import 'list-group';
@import './jfi-header';

:host {
  @include host-base-styles;
  display: block;
}

justifi-payment-method-form {
  display: block;
  margin-left: calc(var(--jfi-layout-padding) * -1);
  margin-right: calc(var(--jfi-layout-padding) * -1);
}

.jfi-checkout-core {
  color: var(--jfi-body-color);
}

*::part(label) {
  color: var(--jfi-form-label-color);
  font-family: var(--jfi-form-label-font-family);
  font-weight: var(--jfi-form-label-font-weight);
  font-size: var(--jfi-form-label-font-size);
  margin: var(--jfi-form-label-margin);
}

*::part(input) {
  font-family: var(--jfi-layout-font-family);
  background-color: var(--jfi-form-control-background-color);
  border-color: var(--jfi-form-control-border-color);
  border-width: var(--jfi-form-control-border-width);
  border-bottom-width: var(--jfi-form-control-border-bottom-width);
  border-left-width: var(--jfi-form-control-border-left-width);
  border-right-width: var(--jfi-form-control-border-right-width);
  border-top-width: var(--jfi-form-control-border-top-width);
  border-radius: var(--jfi-form-control-border-radius);
  border-style: var(--jfi-form-control-border-style);
  box-shadow: var(--jfi-form-control-box-shadow);
  color: var(--jfi-form-control-color);
  font-size: var(--jfi-form-control-font-size);
  font-weight: var(--jfi-form-control-font-weight);
  line-height: var(--jfi-form-control-line-height);
  margin: var(--jfi-form-control-margin);
  padding: var(--jfi-form-control-padding);

  &:focus {
    color: var(--jfi-form-control-color-focus);
    border-color: var(--jfi-form-control-border-color-focus);
    box-shadow: var(--jfi-form-control-box-shadow-focus);
  }
}

*::part(input-invalid) {
  border-color: var(--jfi-form-control-border-color-error);
  box-shadow: var(--jfi-form-control-box-shadow-error);

  &:focus {
    border-color: var(--jfi-form-control-border-color-error);
    box-shadow: var(--jfi-form-control-box-shadow-error-focus);
  }
}

*::part(invalid-feedback) {
  color: var(--jfi-error-message-color);
  font-size: var(--jfi-error-message-font-size);
  margin: var(--jfi-error-message-margin);
  line-height: normal;
}

.jfi-submit-button {
  font-family: var(--jfi-layout-font-family);
  color: var(--jfi-submit-button-color);
  background-color: var(--jfi-submit-button-background-color);
  border-color: var(--jfi-submit-button-border-color);
  padding: var(--jfi-submit-button-padding);
  font-size: var(--jfi-submit-button-font-size);
  border-radius: var(--jfi-submit-button-border-radius);
  width: var(--jfi-submit-button-width);
  box-shadow: var(--jfi-submit-button-box-shadow);
  line-height: var(--jfi-submit-button-line-height);
  text-transform: var(--jfi-submit-button-text-transform);
}

.jfi-submit-button:hover {
  color: var(--jfi-submit-button-color-hover);
  background-color: var(--jfi-submit-button-background-color-hover);
  border-color: var(--jfi-submit-button-border-color-hover);
}

.jfi-submit-button:focus {
  color: var(--jfi-submit-button-color-focus);
  background-color: var(--jfi-submit-button-background-color-focus);
  border-color: var(--jfi-submit-button-border-color-focus);
}

// Extra selector for specificity
.btn.jfi-submit-button:active {
  color: var(--jfi-submit-button-color-active);
  background-color: var(--jfi-submit-button-background-color-active);
  border-color: var(--jfi-submit-button-border-color-active);
}

.btn.jfi-submit-button-loading {
  min-width: 75px;
  color: var(--jfi-submit-button-color-loading);
  background-color: var(--jfi-submit-button-background-color-loading);
  border-color: var(--jfi-submit-button-border-color-loading);
}

.btn.jfi-submit-button-loading .spinner-border {
  border-top-color: var(--jfi-submit-button-color-loading);
  border-bottom-color: var(--jfi-submit-button-color-loading);
  border-left-color: var(--jfi-submit-button-color-loading);
}