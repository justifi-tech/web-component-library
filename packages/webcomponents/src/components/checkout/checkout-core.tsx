import { Component, h, Prop, State, Event, EventEmitter, Host, Method } from '@stencil/core';
import { extractComputedFontsToLoad, formatCurrency } from '../../utils/utils';
import { config } from '../../../config';
import { PaymentMethodPayload } from './payment-method-payload';
import { Checkout, ICheckout, ICheckoutCompleteResponse } from '../../api/Checkout';
import { ComponentError, ComponentErrorCodes, ComponentErrorSeverity } from '../../api/ComponentError';

@Component({
  tag: 'justifi-checkout-core',
  styleUrl: 'checkout-core.scss',
  shadow: true,
})
export class CheckoutCore {
  /**
 * URL for the rendered iFrame. End-users need not use this.
 */
  @Prop({ mutable: true }) iframeOrigin?: string = config.iframeOrigin;
  @Prop() authToken: string;
  @Prop() getCheckout: Function;
  @Prop() complete: Function;
  @Prop() checkoutId: string;

  @Prop() disableCreditCard?: boolean;
  @Prop() disableBankAccount?: boolean;
  @Prop() disableBnpl?: boolean;
  @Prop() disablePaymentMethodGroup?: boolean;

  @State() hasLoadedFonts: boolean = false;
  @State() checkout: ICheckout;
  @State() serverError: string;
  @State() renderState: 'loading' | 'error' | 'success' = 'loading';
  @State() creatingNewPaymentMethod: boolean = false;

  @Event({ eventName: 'submitted' }) submitted: EventEmitter<ICheckoutCompleteResponse>;
  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentError>;

  private paymentMethodOptionsRef?: HTMLJustifiPaymentMethodOptionsElement;

  componentWillLoad() {
    if (this.getCheckout) {
      this.fetchData();
    }
  }

  connectedCallback() {
    if (!this.hasLoadedFonts) {
      this.loadFontsOnParent();
      this.hasLoadedFonts = true;
    }
  }

  fetchData(): void {
    this.renderState = 'loading';
    this.getCheckout({
      onSuccess: ({ checkout }) => {
        this.checkout = new Checkout(checkout);
        this.renderState = 'success';
      },
      onError: ({ error, code, severity }) => {
        this.serverError = error;
        this.renderState = 'error';
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
      }
    });
  };


  @Method()
  async loadFontsOnParent() {
    const parent = document.body;
    const fontsToLoad = extractComputedFontsToLoad();
    if (!parent || !fontsToLoad) {
      return null;
    }

    // This approach is needed to load the font in a parent of the component
    const fonts = document.createElement('link');
    fonts.rel = 'stylesheet';
    fonts.href = `https://fonts.googleapis.com/css2?family=${fontsToLoad}&display=swap`;

    parent.append(fonts);
  }

  async submit(event) {
    event.preventDefault();

    this.renderState = 'loading';

    const payload: PaymentMethodPayload = await this.paymentMethodOptionsRef.resolvePaymentMethod();

    if (!payload) {
      this.renderState = 'error';
    }
    else if (payload.error) {
      this.renderState = 'error';
      this.serverError = payload.error.message;
      this.onError({
        code: (payload.error.code as ComponentErrorCodes),
        error: payload.error.message,
        severity: ComponentErrorSeverity.ERROR,
      });
    }
    else if (payload.token) {
      this.complete({
        payment: { payment_mode: 'ecom', payment_token: payload.token },
        onSuccess: this.onSubmitted,
        onError: this.onError,
      })
    }
    else if (payload.bnpl?.status === 'success') {
      this.complete({
        payment: { payment_mode: 'bnpl' },
        onSuccess: this.onSubmitted,
        onError: this.onError,
      })
    }
  }

  onSubmitted = (data: ICheckoutCompleteResponse) => {
    this.submitted.emit(data);
    this.renderState = 'success';
  };

  onError = ({ error, code, severity }: { error: string, code: ComponentErrorCodes, severity: ComponentErrorSeverity }) => {
    this.serverError = error;
    this.errorEvent.emit({
      errorCode: code,
      message: error,
      severity,
    });
    this.renderState = 'error';
  };

  get isLoading() {
    return this.renderState === 'loading';
  }

  get isError() {
    return this.renderState === 'error';
  }

  get paymentType() {
    if (this.isError) {
      // For now, just return nothing to avoid breaking, but we can decide to show an error message here
      // return <div style={{ color: 'red' }}>Error: {this.serverError}</div>;
      return null;
    }

    if (this.isLoading) {
      return <justifi-skeleton variant="rounded" height="300px" />;
    }

    return (
      <justifi-payment-method-options
        ref={(el) => (this.paymentMethodOptionsRef = el)}
        show-card={!this.disableCreditCard}
        show-ach={!this.disableBankAccount}
        show-bnpl={!this.disableBnpl}
        show-saved-payment-methods={!this.disablePaymentMethodGroup}
        bnpl={this.checkout?.bnpl}
        client-id={this.checkout?.payment_client_id}
        account-id={this.checkout?.account_id}
        savedPaymentMethods={this.checkout?.payment_methods || []}
        paymentAmount={this.checkout?.payment_amount}
      />
    );
  }

  get summary() {
    if (this.isLoading) {
      return <justifi-skeleton height="50px" />;
    }

    if (this.isError) {
      return null;
    }

    return (
      <div>
        <div class="jfi-payment-description">{this.checkout?.payment_description}</div>
        <div class="jfi-payment-total">
          <span class="jfi-payment-total-label">Total</span>&nbsp;
          <span class="jfi-payment-total-amount">{formatCurrency(+this.checkout.payment_amount)}</span>
        </div>
      </div>
    );
  }

  get loadingSpinner() {
    return <justifi-loading-spinner />;
  }

  render() {
    return (
      <Host>
        <div class="row gy-3 jfi-checkout-core">
          <div class="col-12 mb-4">
            {/* componentize this */}
            <h2 class="fs-5 fw-bold pb-3 jfi-header">Summary</h2>
            {this.summary}
          </div>

          <div class="col-12">
            <h2 class="fs-5 fw-bold pb-3 jfi-header">Payment</h2>
            <h3 class="fs-6 fw-bold lh-lg">Select payment type</h3>
            <div class="d-flex flex-column">
              {this.paymentType}
            </div>
          </div>
          <div class="col-12">
            <div class="d-flex justify-content-end">
              <button
                type="submit"
                onClick={event => this.submit(event)}
                disabled={this.isLoading}
                part="pay-button"
                class={`btn btn-primary jfi-submit-button ${this.isLoading ? 'jfi-submit-button-loading' : ''}`}
              >
                {this.isLoading ? this.loadingSpinner : 'Pay'}
              </button>
            </div>
          </div>
        </div>
      </Host>
    );
  }
}
