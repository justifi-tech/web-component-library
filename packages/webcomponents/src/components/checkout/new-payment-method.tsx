import { Component, h, Prop, Method, Event, EventEmitter, Listen, State, Host } from '@stencil/core';
import { config } from '../../../config';
import { PaymentMethodOption } from './payment-method-option-utils';
import { PaymentMethodPayload } from './payment-method-payload';
import { BillingFormFields } from '../billing-form/billing-form-schema';
import { Header3 } from '../../ui-components';
import { billingForm, radioListItem } from '../../styles/parts';

const PaymentMethodTypeLabels = {
  bankAccount: 'New bank account',
  card: 'New credit or debit card',
};

@Component({
  tag: 'justifi-new-payment-method',
})
export class NewPaymentMethod {
  private billingFormRef?: HTMLJustifiBillingFormElement;
  private paymentMethodFormRef?: HTMLCardFormElement;

  @Prop({ mutable: true }) iframeOrigin?: string = config.iframeOrigin;
  @Prop() authToken: string;
  @Prop() accountId: string;
  @Prop() paymentMethodOption: PaymentMethodOption;
  @Prop() paymentMethodGroupId?: string;
  @Prop() isSelected: boolean;

  @State() saveNewPaymentMethodChecked: boolean = false;

  @Event({ bubbles: true }) paymentMethodOptionSelected: EventEmitter;

  @Listen('checkboxChanged')
  handleCheckboxChanged(event: CustomEvent<boolean>) {
    this.saveNewPaymentMethodChecked = event.detail;
  }

  @Method()
  async fillBillingForm(fields: BillingFormFields) {
    this.billingFormRef.fill(fields);
  }

  @Method()
  async resolvePaymentMethod(insuranceValidation: any): Promise<PaymentMethodPayload> {
    if (!this.paymentMethodFormRef || !this.billingFormRef) return;

    try {
      const isValid = await this.validate();

      if (!isValid || !insuranceValidation.isValid) {
        return { validationError: true };
      }

      const tokenizeResponse = await this.tokenize();

      if (tokenizeResponse.error) {
        return { error: tokenizeResponse.error };
      } else {
        const tokenizeRessponseData = tokenizeResponse.data;
        return { token: tokenizeRessponseData.card?.token || tokenizeRessponseData.bank_account?.token };
      }
    } catch (error) {
      return { error };
    }
  }

  async validate(): Promise<boolean> {
    const billingFormValidation = await this.billingFormRef.validate();
    const paymentMethodFormValidation = await this.paymentMethodFormRef.validate();
    return billingFormValidation.isValid && paymentMethodFormValidation.isValid;
  }

  async tokenize() {
    try {
      const billingFormFieldValues = await this.billingFormRef.getValues();
      let paymentMethodData;
      if (this.saveNewPaymentMethodChecked) {
        paymentMethodData = { ...billingFormFieldValues, payment_method_group_id: this.paymentMethodGroupId };
      } else {
        paymentMethodData = { ...billingFormFieldValues };
      }
      const clientId = this.authToken;
      const tokenizeResponse = await this.paymentMethodFormRef.tokenize(clientId, paymentMethodData, this.accountId);
      return tokenizeResponse;
    } catch (error) {
      return error;
    }
  }

  onPaymentMethodOptionClick = () => {
    this.paymentMethodOptionSelected.emit(this.paymentMethodOption);
  };

  showNewPaymentMethodForm() {
    return (
      <div class="mt-4 pb-4 border-bottom">
        <hidden-input />
        <div class="mb-4">
          {this.paymentMethodOption?.id === 'card' ? (
            <card-form ref={(el) => this.paymentMethodFormRef = el} />
          ) : (
            <bank-account-form ref={(el) => this.paymentMethodFormRef = el} />
          )}

        </div>
        <div part={billingForm}>
          <Header3 text="Billing address" class="fs-6 fw-bold lh-lg mb-4" />
          <justifi-billing-form ref={(el) => (this.billingFormRef = el)} />
        </div>
        <justifi-save-new-payment-method hidden={!this.paymentMethodGroupId} />
      </div>
    );
  }

  render() {
    return (
      <Host class="payment-method">
        <div
          class="radio-list-item p-3"
          part={radioListItem}
        >
          <form-control-radio
            id={this.paymentMethodOption?.id}
            name="paymentMethodType"
            value={this.paymentMethodOption?.id}
            checked={this.isSelected}
            label={PaymentMethodTypeLabels[this.paymentMethodOption?.id]}
            inputHandler={this.onPaymentMethodOptionClick}
          />
        </div>

        {this.isSelected ? this.showNewPaymentMethodForm() : null}
      </Host>
    );
  }
}
