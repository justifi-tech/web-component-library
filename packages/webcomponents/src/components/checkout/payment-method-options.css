:host {
  display: block;
}

input[type="checkbox"] {
  cursor: pointer;
}

.checkbox-label {
  cursor: auto;
}

.payment-method-header {
  cursor: pointer;
  color: var(--jfi-radio-button-group-color);
  background-color: var(--jfi-radio-button-group-background-color);
  border-bottom: var(--jfi-radio-button-group-divider);
}

.payment-method-header:hover {
  color: var(--jfi-radio-button-group-color-hover);
  background-color: var(--jfi-radio-button-group-background-color-hover);
}