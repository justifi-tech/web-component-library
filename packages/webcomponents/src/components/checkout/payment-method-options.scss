@import 'forms';
@import 'buttons';
@import 'bootstrap-utilities';

:host {
  @include host-base-styles;
  display: block;
}

.payment-method-header {
  cursor: pointer;

  input,
  label {
    cursor: pointer;
  }

  &:hover {
    background-color: var(--bs-gray-100);
  }
}