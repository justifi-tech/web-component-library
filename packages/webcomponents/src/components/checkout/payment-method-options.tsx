import { Component, Event, EventEmitter, h, Prop, State, Watch, Listen, Method, Host } from '@stencil/core';
import { PaymentMethodTypes } from '../../api/Payment';
import { PaymentMethodOption } from './payment-method-option-utils';
import { PaymentMethodPayload } from './payment-method-payload';
import { IBnpl } from '../../api';
import { BillingFormFields, PostalFormFields } from '../billing-forms/billing-form-schema';

@Component({
  tag: 'justifi-payment-method-options',
})
export class PaymentMethodOptions {
  @Prop() showCard: boolean;
  @Prop() showAch: boolean;
  @Prop() showBnpl: boolean;
  @Prop() showSavedPaymentMethods: boolean;
  @Prop() paymentMethodGroupId?: string;
  @Prop() bnpl: IBnpl;
  @Prop() insuranceToggled: boolean;
  @Prop() authToken: string;
  @Prop() accountId: string;
  @Prop() savedPaymentMethods: any[] = [];
  @Prop() paymentAmount: number;
  @Prop() hideCardBillingForm?: boolean;

  @State() selectedPaymentMethodId: string;
  @State() paymentMethodOptions: PaymentMethodOption[] = [];

  @Event({ bubbles: true }) toggleCreatingNewPaymentMethod: EventEmitter;

  private selectedPaymentMethodOptionRef?: HTMLJustifiNewPaymentMethodElement | HTMLJustifiSavedPaymentMethodElement | HTMLJustifiSezzlePaymentMethodElement;

  @Method()
  async fillBillingForm(fields: BillingFormFields | PostalFormFields) {
    const newPaymentMethodElement = (this.selectedPaymentMethodOptionRef as HTMLJustifiNewPaymentMethodElement);
    if (newPaymentMethodElement.fillBillingForm) {
      newPaymentMethodElement.fillBillingForm(fields);
    }
  }

  connectedCallback() {
    this.paymentMethodsChanged();
  }

  @Watch('savedPaymentMethods')
  paymentMethodsChanged() {
    this.paymentMethodOptions = this.savedPaymentMethods
      .map((paymentMethod) => new PaymentMethodOption(paymentMethod))
      .filter((paymentMethod) => {
        // Don't saved card or bank account if they are disabled
        return (
          (this.showCard || paymentMethod.type !== PaymentMethodTypes.card) &&
          (this.showAch || paymentMethod.type !== PaymentMethodTypes.bankAccount)
        );
      });
    if (this.showBnpl && this.bnpl?.provider === 'sezzle' && !this.insuranceToggled) {
      this.paymentMethodOptions.push(new PaymentMethodOption({ id: PaymentMethodTypes.sezzle }));
    }
    if (this.showCard) {
      this.paymentMethodOptions.push(new PaymentMethodOption({ id: PaymentMethodTypes.card }));
    }
    if (this.showAch) {
      this.paymentMethodOptions.push(new PaymentMethodOption({ id: PaymentMethodTypes.bankAccount }));
    }
    if (!this.selectedPaymentMethodId) {
      this.selectedPaymentMethodId = this.paymentMethodOptions[0]?.id;
    }
  }

  @Listen('paymentMethodOptionSelected')
  paymentMethodOptionSelected(event: CustomEvent<PaymentMethodOption>) {
    this.selectedPaymentMethodId = event.detail.id;
  }

  @Method()
  async resolvePaymentMethod(insuranceValidation: any): Promise<PaymentMethodPayload> {
    return await this.selectedPaymentMethodOptionRef?.resolvePaymentMethod(insuranceValidation);
  }

  render() {
    return (
      <Host>
        {this.paymentMethodOptions?.map((paymentMethodOption) => {
          const newCard = paymentMethodOption.id === PaymentMethodTypes.card;
          const newBankAccount = paymentMethodOption.id === PaymentMethodTypes.bankAccount;
          const isSelected = this.selectedPaymentMethodId === paymentMethodOption.id;
          const sezzle = paymentMethodOption.id === PaymentMethodTypes.sezzle;
          if (newCard || newBankAccount) {
            return (
              <justifi-new-payment-method
                paymentMethodOption={paymentMethodOption}
                authToken={this.authToken}
                account-id={this.accountId}
                is-selected={isSelected}
                paymentMethodGroupId={this.paymentMethodGroupId}
                hideCardBillingForm={this.hideCardBillingForm}
                ref={(el) => {
                  if (isSelected) {
                    this.selectedPaymentMethodOptionRef = el;
                  }
                }}
              />
            );
          } else if (sezzle) {
            return (
              <justifi-sezzle-payment-method
                paymentMethodOption={paymentMethodOption}
                is-selected={isSelected}
                paymentAmount={this.paymentAmount}
                bnpl={this.bnpl}
                ref={(el) => {
                  if (isSelected) {
                    this.selectedPaymentMethodOptionRef = el;
                  }
                }}>
              </justifi-sezzle-payment-method>);
          }
          else if (this.showSavedPaymentMethods) {
            return (
              <justifi-saved-payment-method
                paymentMethodOption={paymentMethodOption}
                is-selected={isSelected}
                ref={(el) => {
                  if (isSelected) {
                    this.selectedPaymentMethodOptionRef = el;
                  }
                }}
              />
            );
          }
        })}
      </Host>
    );
  }
}
