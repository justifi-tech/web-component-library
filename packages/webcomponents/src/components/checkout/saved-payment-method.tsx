import { Component, Event, EventEmitter, h, Prop, Method, Host } from '@stencil/core';
import { config } from '../../../config';
import { CardBrandLabels, PaymentMethodOption } from './payment-method-option-utils';
import { PaymentMethodPayload } from './payment-method-payload';
import { radioListItem } from '../../styles/parts';

@Component({
  tag: 'justifi-saved-payment-method',
})
export class SavedPaymentMethod {
  @Prop({ mutable: true }) iframeOrigin?: string = config.iframeOrigin;
  @Prop() paymentMethodOption: PaymentMethodOption;
  @Prop() isSelected: boolean;

  @Event({ bubbles: true }) paymentMethodOptionSelected: EventEmitter;

  @Method()
  async resolvePaymentMethod(insuranceValidation: any): Promise<PaymentMethodPayload> {
    if (!insuranceValidation.isValid) {
      return { validationError: true };
    }
    return { token: this.paymentMethodOption?.id };
  };

  onPaymentMethodOptionClick = (e) => {
    e.preventDefault();
    this.paymentMethodOptionSelected.emit(this.paymentMethodOption);
  };

  render() {
    return (
      <Host class="payment-method">
        <div
          class="radio-list-item p-3"
          part={radioListItem}
          onClick={this.onPaymentMethodOptionClick}
        >
          <form-control-radio
            name="paymentMethodType"
            value={this.paymentMethodOption?.id}
            checked={this.isSelected}
            label={`${CardBrandLabels[this.paymentMethodOption?.brand] || ''} *${this.paymentMethodOption?.acct_last_four}`}
          />
        </div>
      </Host>
    );
  }
}
