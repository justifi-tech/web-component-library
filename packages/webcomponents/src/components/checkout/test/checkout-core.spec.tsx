import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { CheckoutCore } from '../checkout-core';
import { makeCheckoutComplete, makeGetCheckout } from '../checkout-actions';
import { Checkout, ICheckout } from '../../../api/Checkout';
import { API_NOT_AUTHENTICATED_ERROR } from '../../../api/shared';
import mockGetCheckoutSuccess from './../../../../../../mockData/mockGetCheckoutSuccess.json';
import mockPostCheckoutSuccess from './../../../../../../mockData/mockPostCheckoutSuccess.json';
import { PaymentMethodOptions } from '../payment-method-options';
import { SavedPaymentMethod } from '../saved-payment-method';
import { NewPaymentMethod } from '../new-payment-method';
import { SezzelPaymentMethod } from '../sezzel-payment-method';

describe('justifi-checkout-core', () => {
  it('should display loading state correclty', async () => {
    const page = await newSpecPage({
      components: [CheckoutCore],
      template: () => <justifi-checkout-core />,
    });

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('should display error state correctly', async () => {
    const getCheckout = makeGetCheckout({
      checkoutId: '',
      authToken: '',
      service: {
        fetchCheckout: async () => ({ error: 'error' }),
      },
    });

    const page = await newSpecPage({
      components: [CheckoutCore],
      template: () => <checkout-core getCheckout={getCheckout} />,
    });

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('should set checkout correctly to state', async () => {
    const getCheckout = makeGetCheckout({
      checkoutId: '',
      authToken: '',
      service: {
        fetchCheckout: async () => mockGetCheckoutSuccess,
      },
    });

    const page = await newSpecPage({
      components: [CheckoutCore],
      template: () => <justifi-checkout-core getCheckout={getCheckout} />,
    });

    await page.waitForChanges();

    const expectedCheckout = new Checkout(mockGetCheckoutSuccess.data as unknown as ICheckout);
    expect(page.rootInstance.checkout).toEqual(expectedCheckout);
    expect(page.root).toMatchSnapshot();
  });

  it('should emit error event when fetch fails', async () => {
    const getCheckout = makeGetCheckout({
      authToken: '',
      checkoutId: '',
      service: {
        fetchCheckout: jest.fn().mockRejectedValue(new Error('Fetch error')),
      },
    })

    const errorSpy = jest.fn();

    const page = await newSpecPage({
      components: [CheckoutCore],
      template: () => <justifi-checkout-core getCheckout={getCheckout} onError-event={errorSpy} />,
    });

    await page.waitForChanges();

    expect(errorSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'fetch-error',
          message: 'Fetch error',
          severity: 'error',
        }
      })
    );
  });

  it('should emit error event when API returns error', async () => {
    const getCheckout = makeGetCheckout({
      authToken: '',
      checkoutId: '',
      service: {
        fetchCheckout: jest.fn().mockResolvedValue(API_NOT_AUTHENTICATED_ERROR),
      },
    })

    const errorSpy = jest.fn();

    const page = await newSpecPage({
      components: [CheckoutCore],
      template: () => <justifi-checkout-core getCheckout={getCheckout} onError-event={errorSpy} />,
    });

    await page.waitForChanges();

    expect(errorSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'not-authenticated',
          message: 'Not Authenticated',
          severity: 'error',
        }
      })
    );
  });

  it('emits "submitted" event upon submit request completion', async () => {
    const getCheckout = makeGetCheckout({
      authToken: '',
      checkoutId: '',
      service: {
        fetchCheckout: async () => mockGetCheckoutSuccess,
      }
    });

    const checkoutComplete = makeCheckoutComplete({
      authToken: '',
      checkoutId: '',
      service: {
        fetchCheckout: async () => mockPostCheckoutSuccess,
      }
    });

    const submittedSpy = jest.fn();

    const page = await newSpecPage({
      components: [CheckoutCore, PaymentMethodOptions],
      template: () => <justifi-checkout-core getCheckout={getCheckout} complete={checkoutComplete} onSubmitted={submittedSpy} />,
    });

    await page.waitForChanges();

    // mock resolved payment method (token) / select payment method
    const instance = page.rootInstance;
    instance.paymentMethodOptionsRef = { resolvePaymentMethod: jest.fn().mockResolvedValue({ token: '123' }) };

    // Submit the form
    const submitButton = page.root.shadowRoot.querySelector('button[type="submit"]');
    (submitButton as HTMLElement).click();

    expect(submittedSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          // checkout stuff
        }
      })
    );
  });
});
