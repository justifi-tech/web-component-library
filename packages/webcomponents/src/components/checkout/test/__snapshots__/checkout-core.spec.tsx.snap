// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`justifi-checkout-core should display error state correctly 1`] = `<checkout-core />`;

exports[`justifi-checkout-core should display loading state correctly 1`] = `
<justifi-checkout-core>
  <mock:shadow-root>
    <div class="gy-3 jfi-checkout-core row">
      <div class="col-12 mb-4">
        <h2 class="fs-5 fw-bold jfi-header pb-3">
          Summary
        </h2>
        <justifi-skeleton height="50px"></justifi-skeleton>
      </div>
      <div class="col-12">
        <h2 class="fs-5 fw-bold jfi-header pb-3">
          Payment
        </h2>
        <h3 class="fs-6 fw-bold lh-lg">
          Select payment type
        </h3>
        <div class="d-flex flex-column">
          <justifi-skeleton height="300px" variant="rounded"></justifi-skeleton>
        </div>
      </div>
      <div class="col-12">
        <div class="d-flex justify-content-end">
          <button class="btn btn-primary jfi-submit-button jfi-submit-button-loading" disabled="" part="pay-button" type="submit">
            <justifi-loading-spinner></justifi-loading-spinner>
          </button>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</justifi-checkout-core>
`;

exports[`justifi-checkout-core should set checkout correctly to state 1`] = `
<justifi-checkout-core>
  <mock:shadow-root>
    <div class="gy-3 jfi-checkout-core row">
      <div class="col-12 mb-4">
        <h2 class="fs-5 fw-bold jfi-header pb-3">
          Summary
        </h2>
        <div>
          <div class="jfi-payment-description">
            4x Cases of Celsius Energy Drink
          </div>
          <div class="jfi-payment-total">
            <span class="jfi-payment-total-label">
              Total
            </span>
            <span class="jfi-payment-total-amount">
              $85.92
            </span>
          </div>
        </div>
      </div>
      <div class="col-12">
        <h2 class="fs-5 fw-bold jfi-header pb-3">
          Payment
        </h2>
        <h3 class="fs-6 fw-bold lh-lg">
          Select payment type
        </h3>
        <div class="d-flex flex-column">
          <justifi-payment-method-options account-id="acc_4Et9iXrAARKSZ4KSoiQGTYr" client-id="test_ef88f04afebc3c018de30fd3652d7cee" paymentamount="8592" show-ach="" show-bnpl="" show-card="" show-saved-payment-methods=""></justifi-payment-method-options>
        </div>
      </div>
      <div class="col-12">
        <div class="d-flex justify-content-end">
          <button class="btn btn-primary jfi-submit-button" part="pay-button" type="submit">
            Pay
          </button>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</justifi-checkout-core>
`;
